scikeras/wrappers.py

```diff
@@ -871,192 +875,182 @@ def preprocess_y(y):
-            # y = array([1, 0, 5], [2, 1, 3])
-            # split into lists for multi-output Keras
-            # each will be processesed as a seperate multiclass problem
-            y = np.split(y, y.shape[1], axis=1)
-            model_n_outputs_ = len(y)
-            # encode
-            encoders_ = [LabelEncoder() for _ in range(len(y))]
-            y = [
-                encoder.fit_transform(y_.reshape(-1,) if y_.shape[1] == 1 else y_)
-                for encoder, y_ in zip(encoders_, y)
-            ]
-            classes_ = [encoder.classes_ for encoder in encoders_]
-        else:
-            raise ValueError("Unknown label type: {}".format(target_type_))
-
-        # self.classes_ is kept as an array when n_outputs>1 for compatibility
-        # with ensembles and other meta estimators
-        # which do not support multioutput
-        if len(classes_) == 1:
-            n_classes_ = classes_[0].size
-            classes_ = classes_[0]
+            meta = {
+                "classes_": encoder[1].categories_[0],
+                "n_classes_": len(encoder[1].categories_[0]),
+                "encoders_": [encoder],
+                "model_n_outputs_": 1,
+                "n_outputs_": 1,
+            }
+        elif self.target_type_ == "multiclass-one-hot":
+            # y = array([1, 0], [1, 0]) (usually used with categorical_crossentropy)
+            model_n_outputs_ = 1  # single softmax output expected
             n_outputs_ = 1
-        else:
-            n_classes_ = [class_.shape[0] for class_ in classes_]
-            n_outputs_ = len(n_classes_)
-
-        extra_args.update(
-            {
-                "classes_": classes_,
-                "encoders_": encoders_,
-                "n_outputs_": n_outputs_,
-                "model_n_outputs_": model_n_outputs_,
-                "n_classes_": n_classes_,
-                "target_type_": target_type_,
+            # encode
+            # here we just pass through to allow all Keras use cases
+            # that didn't exist in sklearn to succeed (ex: loss=categorical_crossentropy)
+            if reset:
+                encoder = FunctionTransformer()
+            else:
+                encoder = self.encoders_[0]
+            y = encoder.fit_transform(y)
+            meta = {
+                "classes_": np.arange(0, y.shape[1]),
+                "n_classes_": y.shape[1],
+                "encoders_": [encoder],
+                "model_n_outputs_": 1,
+                "n_outputs_": 1,
             }
-        )
+        else:
+            raise ValueError(
+                f"Unknown label type: {self.target_type_}."  # To match errors used by sklearn
+                "\n\nTo implement support, subclass KerasClassifier and override"
+                " `preprocess_y` and `postprocess_y`."
+                "\n\nSee (TODO: link to docs) for more information."
+            )
 
-        return y, extra_args
+        if reset:
+            self.__dict__.update(meta)
```

Comment:
Wouldn't it be cleaner to set `self.meta_ = meta`? That's be the value that's passed to the `meta` keyword in `model_build_fn` right?

Reply:
Well we want to store them as direct attributes, that's what most scikit-learn estimators do for these kinds of fit-generated properties I think. And I wouldn't want to store them as both.

Reply:
I think the fact that `model_build_fn` accepts a `meta` argument is a strong case for collapsing the metadata into a `BaseWrapper.meta_` dictionary. I'm thinking that this would be appropriate for `*_dtype*`, `*_shape`, `n_classes_`, and `target_type_` because I don't see those in [MLPClassifier][1], [SGDClassifier][2] or [LogisticRegression][3].

> that's what most scikit-learn estimators do for these kinds of fit-generated properties I think

Most Scikit-learn estimators have only set a couple attributes in `fit`. Scikit-learn estimators definitely don't set attributes like `is_fitted_` or `X_dtype_`.

[1]:https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html
[2]:https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html
[3]:https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html

Reply:
We could store it that way, but there are some attributes that Scikit-Learn does use in its estimators, for examples `n_classes_` and `n_outputs_`. Would we split them up, or would we have something like:

```python3
class BaseWrapper(...):
    def __getattr__(self, key):
       if key in self._meta:
          return self._meta[key]
       return super().__getattr__(key)
```